From 662cc520f203aad04418d0c27698a7565aaf4f4a Mon Sep 17 00:00:00 2001
From: Tim Starling <tstarling@wikimedia.org>
Date: Thu, 19 Jun 2014 14:16:31 +1000
Subject: [PATCH] Fix boundary cases of strspn, strcspn and substr_replace

I'm all for refactoring, but a precondition for refactoring is that the
original copies of the code in some way resemble each other. While the
bounds checks for strspn, substr_replace and substr might look
superficially similar, in fact they can't be merged with a single
simple if(strict) to take care of the divergence. The main problem is
that multi-pass normalization of the inputs gives you some interesting
and arguably broken behaviour when there are multiple issues that need
to be simultaneously fixed.

So, unrefactor (expand?) string_substr_check() into its three
consistutent parts, with code in the two new parts based closely on the
Zend source in the current git master. Remove the "strict" parameter,
now implicitly true.

Added a test case which previously failed in multiple ways. In
particular, strspn('','x') returned the wrong thing, which broke the
MediaWiki parser quite badly. The expect file was generated by PHP 5.5
and is identical to the one generated by PHP 5.3.

Submitted on behalf of a third-party: The PHP Group
Source: PHP 5.6.0-dev
License: version 3.01 of the PHP license
---
 hphp/runtime/base/zend-string.cpp        |  33 +-
 hphp/runtime/base/zend-string.h          |   5 +-
 hphp/runtime/ext/ext_string.cpp          |  28 +-
 hphp/test/quick/string-bounds.php        |  33 ++
 hphp/test/quick/string-bounds.php.expect | 499 +++++++++++++++++++++++++++++++
 5 files changed, 588 insertions(+), 10 deletions(-)
 create mode 100644 hphp/test/quick/string-bounds.php
 create mode 100644 hphp/test/quick/string-bounds.php.expect

--- a/hphp/runtime/base/zend-string.cpp
+++ b/hphp/runtime/base/zend-string.cpp
@@ -56,7 +56,7 @@ namespace HPHP {
 ///////////////////////////////////////////////////////////////////////////////
 // helpers
 
-bool string_substr_check(int len, int &f, int &l, bool strict /* = true */) {
+bool string_substr_check(int len, int &f, int &l) {
   if (l < 0 && -l > len) {
     return false;
   } else if (l > len) {
@@ -80,7 +80,7 @@ bool string_substr_check(int len, int &f
       f = 0;
     }
   }
-  if (f > len || (f == len && strict)) {
+  if (f >= len) {
     return false;
   }
 
@@ -480,8 +480,33 @@ String string_replace(const char *s, int
                       const char *replacement, int len_repl) {
   assert(s);
   assert(replacement);
-  if (!string_substr_check(len, start, length, false)) {
-    return empty_string();
+
+  // if "start" position is negative, count start position from the end
+  // of the string
+  if (start < 0) {
+    start = len + start;
+    if (start < 0) {
+      start = 0;
+    }
+  }
+  if (start > len) {
+    start = len;
+  }
+  // if "length" position is negative, set it to the length
+  // needed to stop that many chars from the end of the string
+  if (length < 0) {
+    length = (len - start) + length;
+    if (length < 0) {
+      length = 0;
+    }
+  }
+  // check if length is too large
+  if (length > len || (length < 0 && -length > len)) {
+    length = len;
+  }
+  // check if the length is too large adjusting for non-zero start
+  if (start + length > len) {
+    length = len - start;
   }
 
   String retString(len + len_repl - length, ReserveString);
--- a/hphp/runtime/base/zend-string.h
+++ b/hphp/runtime/base/zend-string.h
@@ -265,11 +265,8 @@ String string_convert_hebrew_string(cons
  * Calculates and adjusts "start" and "length" according to string's length.
  * This function determines how those two parameters are interpreted in varies
  * substr-related functions.
- *
- * The parameter strict controls whether to disallow the empty sub-string
- * after the end.
  */
-bool string_substr_check(int len, int &f, int &l, bool strict = true);
+bool string_substr_check(int len, int &f, int &l);
 
 /**
  * Fills a 256-byte bytemask with input. You can specify a range like 'a..z',
--- a/hphp/runtime/ext/ext_string.cpp
+++ b/hphp/runtime/ext/ext_string.cpp
@@ -1145,6 +1145,30 @@ Variant f_substr_count(const String& hay
   return count;
 }
 
+namespace {
+  bool string_strspn_check(int inputLength, int &start, int &scanLength) {
+    if (start < 0) {
+      start += inputLength;
+      if (start < 0) {
+        start = 0;
+      }
+    } else if (start > inputLength) {
+      return false;
+    }
+
+    if (scanLength < 0) {
+      scanLength += (inputLength - start);
+      if (scanLength < 0) {
+        scanLength = 0;
+      }
+    }
+    if (scanLength > inputLength - start) {
+      scanLength = inputLength - start;
+    }
+    return true;
+  }
+}
+
 Variant f_strspn(const String& str1, const String& str2, int start /* = 0 */,
                  int length /* = 0x7FFFFFFF */) {
   const char *s1 = str1.data();
@@ -1152,7 +1176,7 @@ Variant f_strspn(const String& str1, con
   int s1_len = str1.size();
   int s2_len = str2.size();
 
-  if (!string_substr_check(s1_len, start, length)) {
+  if (!string_strspn_check(s1_len, start, length)) {
     return false;
   }
 
@@ -1171,7 +1195,7 @@ Variant f_strcspn(const String& str1, co
   int s1_len = str1.size();
   int s2_len = str2.size();
 
-  if (!string_substr_check(s1_len, start, length)) {
+  if (!string_strspn_check(s1_len, start, length)) {
     return false;
   }
 
--- /dev/null
+++ b/hphp/test/quick/string-bounds.php
@@ -0,0 +1,33 @@
+<?php
+
+// Boundary case truth table for string functions with start and length
+// parameters similar to substr()
+
+function print_result( $desc, $inLen, $start, $scanLen, $result ) {
+  printf( "%-20s %-6d %-6d %-6d %s\n",
+    $desc, $inLen, $start, $scanLen,
+    $result === false ? 'false' : $result );
+}
+
+printf( "%-20s %-6s %-6s %-6s %s\n", "Test", "in", "start", "len", "result" );
+
+for ( $inLen = 0; $inLen <= 2; $inLen++ ) {
+  $xxx = str_repeat( 'x', $inLen );
+  $abc = substr( "abcdefg", 0, $inLen );
+  for ( $start = -$inLen - 1; $start <= $inLen + 1; $start++ ) {
+    for ( $scanLen = -$inLen - 1; $scanLen <= $inLen + 1; $scanLen++ ) {
+      print_result( "strspn(x,x)", $inLen, $start, $scanLen,
+        strspn( $xxx, 'x', $start, $scanLen ) );
+      print_result( "strspn(x,y)", $inLen, $start, $scanLen,
+        strspn( $xxx, 'y', $start, $scanLen ) );
+      print_result( "strcspn(x,x)", $inLen, $start, $scanLen,
+        strcspn( $xxx, 'x', $start, $scanLen ) );
+      print_result( "strcspn(x,y)", $inLen, $start, $scanLen,
+        strcspn( $xxx, 'y', $start, $scanLen ) );
+      print_result( "substr_replace", $inLen, $start, $scanLen,
+        substr_replace( $xxx, 'y', $start, $scanLen ) );
+      print_result( "substr", $inLen, $start, $scanLen,
+        substr( $abc, $start, $scanLen ) );
+    }
+  }
+}
--- /dev/null
+++ b/hphp/test/quick/string-bounds.php.expect
@@ -0,0 +1,499 @@
+Test                 in     start  len    result
+strspn(x,x)          0      -1     -1     0
+strspn(x,y)          0      -1     -1     0
+strcspn(x,x)         0      -1     -1     0
+strcspn(x,y)         0      -1     -1     0
+substr_replace       0      -1     -1     y
+substr               0      -1     -1     false
+strspn(x,x)          0      -1     0      0
+strspn(x,y)          0      -1     0      0
+strcspn(x,x)         0      -1     0      0
+strcspn(x,y)         0      -1     0      0
+substr_replace       0      -1     0      y
+substr               0      -1     0      false
+strspn(x,x)          0      -1     1      0
+strspn(x,y)          0      -1     1      0
+strcspn(x,x)         0      -1     1      0
+strcspn(x,y)         0      -1     1      0
+substr_replace       0      -1     1      y
+substr               0      -1     1      false
+strspn(x,x)          0      0      -1     0
+strspn(x,y)          0      0      -1     0
+strcspn(x,x)         0      0      -1     0
+strcspn(x,y)         0      0      -1     0
+substr_replace       0      0      -1     y
+substr               0      0      -1     false
+strspn(x,x)          0      0      0      0
+strspn(x,y)          0      0      0      0
+strcspn(x,x)         0      0      0      0
+strcspn(x,y)         0      0      0      0
+substr_replace       0      0      0      y
+substr               0      0      0      false
+strspn(x,x)          0      0      1      0
+strspn(x,y)          0      0      1      0
+strcspn(x,x)         0      0      1      0
+strcspn(x,y)         0      0      1      0
+substr_replace       0      0      1      y
+substr               0      0      1      false
+strspn(x,x)          0      1      -1     false
+strspn(x,y)          0      1      -1     false
+strcspn(x,x)         0      1      -1     false
+strcspn(x,y)         0      1      -1     false
+substr_replace       0      1      -1     y
+substr               0      1      -1     false
+strspn(x,x)          0      1      0      false
+strspn(x,y)          0      1      0      false
+strcspn(x,x)         0      1      0      false
+strcspn(x,y)         0      1      0      false
+substr_replace       0      1      0      y
+substr               0      1      0      false
+strspn(x,x)          0      1      1      false
+strspn(x,y)          0      1      1      false
+strcspn(x,x)         0      1      1      false
+strcspn(x,y)         0      1      1      false
+substr_replace       0      1      1      y
+substr               0      1      1      false
+strspn(x,x)          1      -2     -2     0
+strspn(x,y)          1      -2     -2     0
+strcspn(x,x)         1      -2     -2     0
+strcspn(x,y)         1      -2     -2     0
+substr_replace       1      -2     -2     yx
+substr               1      -2     -2     false
+strspn(x,x)          1      -2     -1     0
+strspn(x,y)          1      -2     -1     0
+strcspn(x,x)         1      -2     -1     0
+strcspn(x,y)         1      -2     -1     0
+substr_replace       1      -2     -1     yx
+substr               1      -2     -1     
+strspn(x,x)          1      -2     0      0
+strspn(x,y)          1      -2     0      0
+strcspn(x,x)         1      -2     0      0
+strcspn(x,y)         1      -2     0      0
+substr_replace       1      -2     0      yx
+substr               1      -2     0      
+strspn(x,x)          1      -2     1      1
+strspn(x,y)          1      -2     1      0
+strcspn(x,x)         1      -2     1      0
+strcspn(x,y)         1      -2     1      1
+substr_replace       1      -2     1      y
+substr               1      -2     1      a
+strspn(x,x)          1      -2     2      1
+strspn(x,y)          1      -2     2      0
+strcspn(x,x)         1      -2     2      0
+strcspn(x,y)         1      -2     2      1
+substr_replace       1      -2     2      y
+substr               1      -2     2      a
+strspn(x,x)          1      -1     -2     0
+strspn(x,y)          1      -1     -2     0
+strcspn(x,x)         1      -1     -2     0
+strcspn(x,y)         1      -1     -2     0
+substr_replace       1      -1     -2     yx
+substr               1      -1     -2     false
+strspn(x,x)          1      -1     -1     0
+strspn(x,y)          1      -1     -1     0
+strcspn(x,x)         1      -1     -1     0
+strcspn(x,y)         1      -1     -1     0
+substr_replace       1      -1     -1     yx
+substr               1      -1     -1     
+strspn(x,x)          1      -1     0      0
+strspn(x,y)          1      -1     0      0
+strcspn(x,x)         1      -1     0      0
+strcspn(x,y)         1      -1     0      0
+substr_replace       1      -1     0      yx
+substr               1      -1     0      
+strspn(x,x)          1      -1     1      1
+strspn(x,y)          1      -1     1      0
+strcspn(x,x)         1      -1     1      0
+strcspn(x,y)         1      -1     1      1
+substr_replace       1      -1     1      y
+substr               1      -1     1      a
+strspn(x,x)          1      -1     2      1
+strspn(x,y)          1      -1     2      0
+strcspn(x,x)         1      -1     2      0
+strcspn(x,y)         1      -1     2      1
+substr_replace       1      -1     2      y
+substr               1      -1     2      a
+strspn(x,x)          1      0      -2     0
+strspn(x,y)          1      0      -2     0
+strcspn(x,x)         1      0      -2     0
+strcspn(x,y)         1      0      -2     0
+substr_replace       1      0      -2     yx
+substr               1      0      -2     false
+strspn(x,x)          1      0      -1     0
+strspn(x,y)          1      0      -1     0
+strcspn(x,x)         1      0      -1     0
+strcspn(x,y)         1      0      -1     0
+substr_replace       1      0      -1     yx
+substr               1      0      -1     
+strspn(x,x)          1      0      0      0
+strspn(x,y)          1      0      0      0
+strcspn(x,x)         1      0      0      0
+strcspn(x,y)         1      0      0      0
+substr_replace       1      0      0      yx
+substr               1      0      0      
+strspn(x,x)          1      0      1      1
+strspn(x,y)          1      0      1      0
+strcspn(x,x)         1      0      1      0
+strcspn(x,y)         1      0      1      1
+substr_replace       1      0      1      y
+substr               1      0      1      a
+strspn(x,x)          1      0      2      1
+strspn(x,y)          1      0      2      0
+strcspn(x,x)         1      0      2      0
+strcspn(x,y)         1      0      2      1
+substr_replace       1      0      2      y
+substr               1      0      2      a
+strspn(x,x)          1      1      -2     0
+strspn(x,y)          1      1      -2     0
+strcspn(x,x)         1      1      -2     0
+strcspn(x,y)         1      1      -2     0
+substr_replace       1      1      -2     xy
+substr               1      1      -2     false
+strspn(x,x)          1      1      -1     0
+strspn(x,y)          1      1      -1     0
+strcspn(x,x)         1      1      -1     0
+strcspn(x,y)         1      1      -1     0
+substr_replace       1      1      -1     xy
+substr               1      1      -1     false
+strspn(x,x)          1      1      0      0
+strspn(x,y)          1      1      0      0
+strcspn(x,x)         1      1      0      0
+strcspn(x,y)         1      1      0      0
+substr_replace       1      1      0      xy
+substr               1      1      0      false
+strspn(x,x)          1      1      1      0
+strspn(x,y)          1      1      1      0
+strcspn(x,x)         1      1      1      0
+strcspn(x,y)         1      1      1      0
+substr_replace       1      1      1      xy
+substr               1      1      1      false
+strspn(x,x)          1      1      2      0
+strspn(x,y)          1      1      2      0
+strcspn(x,x)         1      1      2      0
+strcspn(x,y)         1      1      2      0
+substr_replace       1      1      2      xy
+substr               1      1      2      false
+strspn(x,x)          1      2      -2     false
+strspn(x,y)          1      2      -2     false
+strcspn(x,x)         1      2      -2     false
+strcspn(x,y)         1      2      -2     false
+substr_replace       1      2      -2     xy
+substr               1      2      -2     false
+strspn(x,x)          1      2      -1     false
+strspn(x,y)          1      2      -1     false
+strcspn(x,x)         1      2      -1     false
+strcspn(x,y)         1      2      -1     false
+substr_replace       1      2      -1     xy
+substr               1      2      -1     false
+strspn(x,x)          1      2      0      false
+strspn(x,y)          1      2      0      false
+strcspn(x,x)         1      2      0      false
+strcspn(x,y)         1      2      0      false
+substr_replace       1      2      0      xy
+substr               1      2      0      false
+strspn(x,x)          1      2      1      false
+strspn(x,y)          1      2      1      false
+strcspn(x,x)         1      2      1      false
+strcspn(x,y)         1      2      1      false
+substr_replace       1      2      1      xy
+substr               1      2      1      false
+strspn(x,x)          1      2      2      false
+strspn(x,y)          1      2      2      false
+strcspn(x,x)         1      2      2      false
+strcspn(x,y)         1      2      2      false
+substr_replace       1      2      2      xy
+substr               1      2      2      false
+strspn(x,x)          2      -3     -3     0
+strspn(x,y)          2      -3     -3     0
+strcspn(x,x)         2      -3     -3     0
+strcspn(x,y)         2      -3     -3     0
+substr_replace       2      -3     -3     yxx
+substr               2      -3     -3     false
+strspn(x,x)          2      -3     -2     0
+strspn(x,y)          2      -3     -2     0
+strcspn(x,x)         2      -3     -2     0
+strcspn(x,y)         2      -3     -2     0
+substr_replace       2      -3     -2     yxx
+substr               2      -3     -2     
+strspn(x,x)          2      -3     -1     1
+strspn(x,y)          2      -3     -1     0
+strcspn(x,x)         2      -3     -1     0
+strcspn(x,y)         2      -3     -1     1
+substr_replace       2      -3     -1     yx
+substr               2      -3     -1     a
+strspn(x,x)          2      -3     0      0
+strspn(x,y)          2      -3     0      0
+strcspn(x,x)         2      -3     0      0
+strcspn(x,y)         2      -3     0      0
+substr_replace       2      -3     0      yxx
+substr               2      -3     0      
+strspn(x,x)          2      -3     1      1
+strspn(x,y)          2      -3     1      0
+strcspn(x,x)         2      -3     1      0
+strcspn(x,y)         2      -3     1      1
+substr_replace       2      -3     1      yx
+substr               2      -3     1      a
+strspn(x,x)          2      -3     2      2
+strspn(x,y)          2      -3     2      0
+strcspn(x,x)         2      -3     2      0
+strcspn(x,y)         2      -3     2      2
+substr_replace       2      -3     2      y
+substr               2      -3     2      ab
+strspn(x,x)          2      -3     3      2
+strspn(x,y)          2      -3     3      0
+strcspn(x,x)         2      -3     3      0
+strcspn(x,y)         2      -3     3      2
+substr_replace       2      -3     3      y
+substr               2      -3     3      ab
+strspn(x,x)          2      -2     -3     0
+strspn(x,y)          2      -2     -3     0
+strcspn(x,x)         2      -2     -3     0
+strcspn(x,y)         2      -2     -3     0
+substr_replace       2      -2     -3     yxx
+substr               2      -2     -3     false
+strspn(x,x)          2      -2     -2     0
+strspn(x,y)          2      -2     -2     0
+strcspn(x,x)         2      -2     -2     0
+strcspn(x,y)         2      -2     -2     0
+substr_replace       2      -2     -2     yxx
+substr               2      -2     -2     
+strspn(x,x)          2      -2     -1     1
+strspn(x,y)          2      -2     -1     0
+strcspn(x,x)         2      -2     -1     0
+strcspn(x,y)         2      -2     -1     1
+substr_replace       2      -2     -1     yx
+substr               2      -2     -1     a
+strspn(x,x)          2      -2     0      0
+strspn(x,y)          2      -2     0      0
+strcspn(x,x)         2      -2     0      0
+strcspn(x,y)         2      -2     0      0
+substr_replace       2      -2     0      yxx
+substr               2      -2     0      
+strspn(x,x)          2      -2     1      1
+strspn(x,y)          2      -2     1      0
+strcspn(x,x)         2      -2     1      0
+strcspn(x,y)         2      -2     1      1
+substr_replace       2      -2     1      yx
+substr               2      -2     1      a
+strspn(x,x)          2      -2     2      2
+strspn(x,y)          2      -2     2      0
+strcspn(x,x)         2      -2     2      0
+strcspn(x,y)         2      -2     2      2
+substr_replace       2      -2     2      y
+substr               2      -2     2      ab
+strspn(x,x)          2      -2     3      2
+strspn(x,y)          2      -2     3      0
+strcspn(x,x)         2      -2     3      0
+strcspn(x,y)         2      -2     3      2
+substr_replace       2      -2     3      y
+substr               2      -2     3      ab
+strspn(x,x)          2      -1     -3     0
+strspn(x,y)          2      -1     -3     0
+strcspn(x,x)         2      -1     -3     0
+strcspn(x,y)         2      -1     -3     0
+substr_replace       2      -1     -3     xyx
+substr               2      -1     -3     false
+strspn(x,x)          2      -1     -2     0
+strspn(x,y)          2      -1     -2     0
+strcspn(x,x)         2      -1     -2     0
+strcspn(x,y)         2      -1     -2     0
+substr_replace       2      -1     -2     xyx
+substr               2      -1     -2     
+strspn(x,x)          2      -1     -1     0
+strspn(x,y)          2      -1     -1     0
+strcspn(x,x)         2      -1     -1     0
+strcspn(x,y)         2      -1     -1     0
+substr_replace       2      -1     -1     xyx
+substr               2      -1     -1     
+strspn(x,x)          2      -1     0      0
+strspn(x,y)          2      -1     0      0
+strcspn(x,x)         2      -1     0      0
+strcspn(x,y)         2      -1     0      0
+substr_replace       2      -1     0      xyx
+substr               2      -1     0      
+strspn(x,x)          2      -1     1      1
+strspn(x,y)          2      -1     1      0
+strcspn(x,x)         2      -1     1      0
+strcspn(x,y)         2      -1     1      1
+substr_replace       2      -1     1      xy
+substr               2      -1     1      b
+strspn(x,x)          2      -1     2      1
+strspn(x,y)          2      -1     2      0
+strcspn(x,x)         2      -1     2      0
+strcspn(x,y)         2      -1     2      1
+substr_replace       2      -1     2      xy
+substr               2      -1     2      b
+strspn(x,x)          2      -1     3      1
+strspn(x,y)          2      -1     3      0
+strcspn(x,x)         2      -1     3      0
+strcspn(x,y)         2      -1     3      1
+substr_replace       2      -1     3      xy
+substr               2      -1     3      b
+strspn(x,x)          2      0      -3     0
+strspn(x,y)          2      0      -3     0
+strcspn(x,x)         2      0      -3     0
+strcspn(x,y)         2      0      -3     0
+substr_replace       2      0      -3     yxx
+substr               2      0      -3     false
+strspn(x,x)          2      0      -2     0
+strspn(x,y)          2      0      -2     0
+strcspn(x,x)         2      0      -2     0
+strcspn(x,y)         2      0      -2     0
+substr_replace       2      0      -2     yxx
+substr               2      0      -2     
+strspn(x,x)          2      0      -1     1
+strspn(x,y)          2      0      -1     0
+strcspn(x,x)         2      0      -1     0
+strcspn(x,y)         2      0      -1     1
+substr_replace       2      0      -1     yx
+substr               2      0      -1     a
+strspn(x,x)          2      0      0      0
+strspn(x,y)          2      0      0      0
+strcspn(x,x)         2      0      0      0
+strcspn(x,y)         2      0      0      0
+substr_replace       2      0      0      yxx
+substr               2      0      0      
+strspn(x,x)          2      0      1      1
+strspn(x,y)          2      0      1      0
+strcspn(x,x)         2      0      1      0
+strcspn(x,y)         2      0      1      1
+substr_replace       2      0      1      yx
+substr               2      0      1      a
+strspn(x,x)          2      0      2      2
+strspn(x,y)          2      0      2      0
+strcspn(x,x)         2      0      2      0
+strcspn(x,y)         2      0      2      2
+substr_replace       2      0      2      y
+substr               2      0      2      ab
+strspn(x,x)          2      0      3      2
+strspn(x,y)          2      0      3      0
+strcspn(x,x)         2      0      3      0
+strcspn(x,y)         2      0      3      2
+substr_replace       2      0      3      y
+substr               2      0      3      ab
+strspn(x,x)          2      1      -3     0
+strspn(x,y)          2      1      -3     0
+strcspn(x,x)         2      1      -3     0
+strcspn(x,y)         2      1      -3     0
+substr_replace       2      1      -3     xyx
+substr               2      1      -3     false
+strspn(x,x)          2      1      -2     0
+strspn(x,y)          2      1      -2     0
+strcspn(x,x)         2      1      -2     0
+strcspn(x,y)         2      1      -2     0
+substr_replace       2      1      -2     xyx
+substr               2      1      -2     false
+strspn(x,x)          2      1      -1     0
+strspn(x,y)          2      1      -1     0
+strcspn(x,x)         2      1      -1     0
+strcspn(x,y)         2      1      -1     0
+substr_replace       2      1      -1     xyx
+substr               2      1      -1     
+strspn(x,x)          2      1      0      0
+strspn(x,y)          2      1      0      0
+strcspn(x,x)         2      1      0      0
+strcspn(x,y)         2      1      0      0
+substr_replace       2      1      0      xyx
+substr               2      1      0      
+strspn(x,x)          2      1      1      1
+strspn(x,y)          2      1      1      0
+strcspn(x,x)         2      1      1      0
+strcspn(x,y)         2      1      1      1
+substr_replace       2      1      1      xy
+substr               2      1      1      b
+strspn(x,x)          2      1      2      1
+strspn(x,y)          2      1      2      0
+strcspn(x,x)         2      1      2      0
+strcspn(x,y)         2      1      2      1
+substr_replace       2      1      2      xy
+substr               2      1      2      b
+strspn(x,x)          2      1      3      1
+strspn(x,y)          2      1      3      0
+strcspn(x,x)         2      1      3      0
+strcspn(x,y)         2      1      3      1
+substr_replace       2      1      3      xy
+substr               2      1      3      b
+strspn(x,x)          2      2      -3     0
+strspn(x,y)          2      2      -3     0
+strcspn(x,x)         2      2      -3     0
+strcspn(x,y)         2      2      -3     0
+substr_replace       2      2      -3     xxy
+substr               2      2      -3     false
+strspn(x,x)          2      2      -2     0
+strspn(x,y)          2      2      -2     0
+strcspn(x,x)         2      2      -2     0
+strcspn(x,y)         2      2      -2     0
+substr_replace       2      2      -2     xxy
+substr               2      2      -2     false
+strspn(x,x)          2      2      -1     0
+strspn(x,y)          2      2      -1     0
+strcspn(x,x)         2      2      -1     0
+strcspn(x,y)         2      2      -1     0
+substr_replace       2      2      -1     xxy
+substr               2      2      -1     false
+strspn(x,x)          2      2      0      0
+strspn(x,y)          2      2      0      0
+strcspn(x,x)         2      2      0      0
+strcspn(x,y)         2      2      0      0
+substr_replace       2      2      0      xxy
+substr               2      2      0      false
+strspn(x,x)          2      2      1      0
+strspn(x,y)          2      2      1      0
+strcspn(x,x)         2      2      1      0
+strcspn(x,y)         2      2      1      0
+substr_replace       2      2      1      xxy
+substr               2      2      1      false
+strspn(x,x)          2      2      2      0
+strspn(x,y)          2      2      2      0
+strcspn(x,x)         2      2      2      0
+strcspn(x,y)         2      2      2      0
+substr_replace       2      2      2      xxy
+substr               2      2      2      false
+strspn(x,x)          2      2      3      0
+strspn(x,y)          2      2      3      0
+strcspn(x,x)         2      2      3      0
+strcspn(x,y)         2      2      3      0
+substr_replace       2      2      3      xxy
+substr               2      2      3      false
+strspn(x,x)          2      3      -3     false
+strspn(x,y)          2      3      -3     false
+strcspn(x,x)         2      3      -3     false
+strcspn(x,y)         2      3      -3     false
+substr_replace       2      3      -3     xxy
+substr               2      3      -3     false
+strspn(x,x)          2      3      -2     false
+strspn(x,y)          2      3      -2     false
+strcspn(x,x)         2      3      -2     false
+strcspn(x,y)         2      3      -2     false
+substr_replace       2      3      -2     xxy
+substr               2      3      -2     false
+strspn(x,x)          2      3      -1     false
+strspn(x,y)          2      3      -1     false
+strcspn(x,x)         2      3      -1     false
+strcspn(x,y)         2      3      -1     false
+substr_replace       2      3      -1     xxy
+substr               2      3      -1     false
+strspn(x,x)          2      3      0      false
+strspn(x,y)          2      3      0      false
+strcspn(x,x)         2      3      0      false
+strcspn(x,y)         2      3      0      false
+substr_replace       2      3      0      xxy
+substr               2      3      0      false
+strspn(x,x)          2      3      1      false
+strspn(x,y)          2      3      1      false
+strcspn(x,x)         2      3      1      false
+strcspn(x,y)         2      3      1      false
+substr_replace       2      3      1      xxy
+substr               2      3      1      false
+strspn(x,x)          2      3      2      false
+strspn(x,y)          2      3      2      false
+strcspn(x,x)         2      3      2      false
+strcspn(x,y)         2      3      2      false
+substr_replace       2      3      2      xxy
+substr               2      3      2      false
+strspn(x,x)          2      3      3      false
+strspn(x,y)          2      3      3      false
+strcspn(x,x)         2      3      3      false
+strcspn(x,y)         2      3      3      false
+substr_replace       2      3      3      xxy
+substr               2      3      3      false
